app/controllers/incomeSources/add/BusinessAddedObligationsController.scala

```diff
@@ -62,28 +66,40 @@ class BusinessAddedObligationsController @Inject()(authenticate: AuthenticationP
       }
   }
 
-  private def getBackUrl(incomeSourceId: String, isAgent: Boolean): String = {
+  private def getBackUrl(isAgent: Boolean): String = {
     val baseRoute = if (isAgent) controllers.incomeSources.add.routes.BusinessReportingMethodController.showAgent _ else
       controllers.incomeSources.add.routes.BusinessReportingMethodController.show _
-    baseRoute(incomeSourceId).url
+    baseRoute().url
   }
 
-  private def handleRequest(isAgent: Boolean, incomeSourceId: String)(implicit user: MtdItUser[_], ec: ExecutionContext): Future[Result] = {
+  private def handleRequest(isAgent: Boolean)(implicit user: MtdItUser[_], ec: ExecutionContext): Future[Result] = {
     withIncomeSourcesFS {
-      val businessNameAndStartDate = getBusinessNameAndStartDate(incomeSourceId)
-
-      businessNameAndStartDate match {
-        case Some((businessName, startDate)) =>
-          val showPreviousTaxYears: Boolean = startDate.isBefore(dateService.getCurrentTaxYearStart())
-          nextUpdatesService.getObligationsViewModel(incomeSourceId, showPreviousTaxYears) map { viewModel =>
-            val backUrl = getBackUrl(incomeSourceId, isAgent)
-            Ok(obligationsView(businessName = Some(businessName), sources = viewModel, backUrl = backUrl, isAgent = isAgent, incomeSourceType = SelfEmployment))
+      sessionService.get(SessionKeys.incomeSourceId).flatMap {
```

Comment:
this method need a lot of refactoring
1. Consider changing next method to signature below:

`def getBusinessNameAndStartDate(incomeSource: Either[Throwable, Option[String]])(implicit user: MtdItUser[_]): Either[Throwable, (String, LocalDate)]`

2.  ObligtionView model: we need to add businessName, not sure why this is not in the model in the first place.

Here is example how you can refactor this method:

```
  private def handleRequest(isAgent: Boolean)(implicit user: MtdItUser[_], ec: ExecutionContext): Future[Result] = {
    withIncomeSourcesFS {
      // TODO: consider moving next section into separate private method
      val obligationViewModelFuture = for {
        incomeSourceIdMayBe <- sessionService.get(SessionKeys.incomeSourceId)
        res <- Future {
          getBusinessNameAndStartDate(incomeSourceIdMayBe)
        }
        viewModelRes <- res match {
          case Right((businessName, startDate)) =>
            val showPreviousTaxYears: Boolean = startDate.isBefore(dateService.getCurrentTaxYearStart())
            nextUpdatesService.getObligationsViewModel(incomeSourceId, showPreviousTaxYears).map(x => Right(x))
          case Left(ex) =>
            Future {
              Left(ex)
            }
        }
      } yield viewModelRes
      // TODO: + add record on the future as well
      // TODO: 2 / add businessName to the ObligationsViewModel
      obligationViewModelFuture.flatMap(res =>
        res match {
          case Right(obligationViewModel) =>
            val backUrl = getBackUrl(isAgent)
            Future {
              Ok(
                obligationsView(businessName = Some("businessName"), sources = obligationViewModel, backUrl = backUrl, isAgent = isAgent, incomeSourceType = SelfEmployment)
              )
            }

          case Left(ex) =>
            val errorMessage = s"Unable to find incomeSource by id: $incomeSourceId"
            Logger("application").error(errorMessage)
            val errorHandler = if (isAgent) itvcErrorHandlerAgent else itvcErrorHandler
            Future.successful(errorHandler.showInternalServerError())
        }
      )
    }
  }
```